### PR TITLE
implement rand(::BigFloat) (close #13948)

### LINKED ISSUE #13948: rand(::BigFloat)
We once had a `rand!` for BigFloats using mpfr_urandom, see #4845.
Apparently this got removed/got lost. As said in #9836 "probably not high priority other than for completeness". I ran into this when calling iterative algorithms parametrized by element-type T which starting from a random vector  rand(T,n).

### PR DESCRIPTION
**Pull request by rfourquet**
This uses a similar method as for `Float64`: generate uniformly sufficiently many random bits to get a number `x` in `[1, 2)`.
Then there are two simple possibilities (cf. this nice [write-up](https://github.com/JuliaLang/julia/issues/16344#issuecomment-219001816)):
1) return `x-1.0` : then one bit of precision is lost
2) return `rand(Bool) ? x/2 : x/2-0.5`

For `Float64`, we do option 1) as dSFMT generates only 52 random bits per iteration, when `Float64` has 53 bits of precision, and this is way faster (the cost of the branch of option 2 is non-negligible).
For `BigFloat`, the two options have similar performances (within few percents, and option 2) even is slightly faster than 1) for higher precisions), so I would favor this one. On the other hand, option 1) has the advantage that it would behave similarly to other floats when their precision is set to match (e.g. `rand(BigFloat)` would behave like `rand(Float64)` when `precision(BigFloat) == 53`). "RFC" until this is decided (option 1 is commented out here, if someone wants to play with it).

One possibility would be to have something along the lines of having `rand(CloseOpen{T})` generate 1 random a float of type `T` with 1 bit less than the precision, `rand(CloseOpenFull{T})` generate at full precision number, and `rand(T)` generate the best of the 2 options (1 for `Float64` and 2 for `BigFloat`). This would incidentaly solve #16344, by allowing to get random `Float64` at full precision (opt-in), which is not terribly slow after all: on my machine, this takes 14ns for full precision vs 5ns for 52 bits precision.

### CODE REVIEW

File: base/random.jl [hunk 1]
```diff
@@ -475,7 +482,11 @@ end
 
 rand!(A::AbstractArray, ::Type{X}) where {X} = rand!(GLOBAL_RNG, A, X)
 
-function rand!(r::AbstractRNG, A::AbstractArray, s::Union{Dict,Set,IntSet})
+rand!(r::AbstractRNG, A::AbstractArray, ::Type{T}) where {T<:AbstractFloat} =
+    rand!(r, A, CloseOpen{T}())
+rand!(A::AbstractArray, ::Type{T}) where {T<:AbstractFloat} = rand!(GLOBAL_RNG, A, T)
```
Comment (tkelman) [inline base/random.jl]: isn't this completely redundant with the more general version just above?

Comment (rfourquet) [inline base/random.jl]: yes thanks

**Comment by mschauer** [comment]
Isn't the branch optimized away if you do
x/2 - rand(Bool)*0.5 ?

**Comment by rfourquet** [comment]
> Isn't the branch optimized away if you do `x/2 - rand(Bool)*0.5` ?

Excellent, not directly (at least I don't see it) but appartently yes with `0.5*(x-rand(Bool))`, which costs 8ns vs 5ns for `rand()`. I had tried with `ifelse` to avoid the branch, but had not got this speedup. Then it seems worthwhile to offer this full precision `rand(Float)`. 

For `BigFloat`, this is a non-issue.

**Comment by rfourquet** [comment]
I decided to implement the full precision `rand(BigFloat)` by default, and to defer having an API to select full/not-full precision to another PR. I will merge in a couple of days if no objection.

File: doc/src/stdlib/numbers.md [hunk 2]
```diff
@@ -143,12 +143,12 @@ dimension specifications `dims...` (which can be given as a tuple) to generate a
 values.
 
 A `MersenneTwister` or `RandomDevice` RNG can generate random numbers of the following types:
-[`Float16`](@ref), [`Float32`](@ref), [`Float64`](@ref), [`Bool`](@ref), [`Int8`](@ref),
-[`UInt8`](@ref), [`Int16`](@ref), [`UInt16`](@ref), [`Int32`](@ref), [`UInt32`](@ref),
-[`Int64`](@ref), [`UInt64`](@ref), [`Int128`](@ref), [`UInt128`](@ref), [`BigInt`](@ref)
-(or complex numbers of those types). Random floating point numbers are generated uniformly
-in ``[0, 1)``. As `BigInt` represents unbounded integers, the interval must be specified
-(e.g. `rand(big(1:6))`).
+[`Float16`](@ref), [`Float32`](@ref), [`Float64`](@ref), [`Float64`](@ref),[`Bool`](@ref),
```
Comment (tkelman) [inline doc/src/stdlib/numbers.md]: is the Float64 you added here supposed to be BigFloat?

Comment (rfourquet) [inline doc/src/stdlib/numbers.md]: oups yes! (rebase mistake)

**Comment by rfourquet** [comment]
The Travis and circleci failures have to do `RowVector` (inferred return type), it's an already seen failure, and the appveyor failure has to do with "codegen"; so looks unrelated.

**Pull request merged**